tmva/tmva/test/DNN/TestMethodDLOptimization.h

```diff
@@ -0,0 +1,159 @@
+ **********************************************************************************/
+
+#ifndef TMVA_TEST_DNN_TEST_METHOD_DL_OPTIMIZATION_H
+#define TMVA_TEST_DNN_TEST_METHOD_DL_OPTIMIZATION_H
+
+#include "TFile.h"
+#include "TTree.h"
+#include "TString.h"
+#include "TROOT.h"
+
+#include "TMVA/MethodDL.h"
+#include "TMVA/DataLoader.h"
+#include "TMVA/Factory.h"
+#include "TMVA/Config.h"
+
+#include <iostream>
+
+/** Testing the entire pipeline of the Method DL, when only a Multilayer Percepton
+ *  is constructed with various optimizers. */
+//______________________________________________________________________________
+void testMethodDL_DNN(TString architectureStr, TString optimizerStr)
+{
+
+   ROOT::EnableImplicitMT(1);
+   TMVA::Config::Instance();
+
+   TFile *input(0);
+   // TString fname = "tmva_class_example.root";
+   TString fname = "http://root.cern.ch/files/tmva_class_example.root";
+   TString fopt = "CACHEREAD";
+   input = TFile::Open(fname, fopt);
+
+   // Register the training and test trees
+   TTree *signalTree = (TTree *)input->Get("TreeS");
+   TTree *background = (TTree *)input->Get("TreeB");
+
+   TString outfileName("TMVA_DNN.root");
+   TFile *outputFile = TFile::Open(outfileName, "RECREATE");
+
+   TMVA::DataLoader *dataloader = new TMVA::DataLoader("dataset");
+
+   dataloader->AddVariable("myvar1 := var1+var2", 'F');
+   dataloader->AddVariable("myvar2 := var1-var2", "Expression 2", "", 'F');
+   dataloader->AddVariable("var3", "Variable 3", "units", 'F');
+   dataloader->AddVariable("var4", "Variable 4", "units", 'F');
+   dataloader->AddSpectator("spec1 := var1*2", "Spectator 1", "units", 'F');
+   dataloader->AddSpectator("spec2 := var1*3", "Spectator 2", "units", 'F');
+
+   // global event weights per tree
+   Double_t signalWeight = 1.0;
+   Double_t backgroundWeight = 1.0;
+
+   // Add signal and background trees
+   dataloader->AddSignalTree(signalTree, signalWeight);
+   dataloader->AddBackgroundTree(background, backgroundWeight);
+
+   dataloader->SetBackgroundWeightExpression("weight");
+
+   TCut mycuts = "";
+   TCut mycutb = "";
+
+   dataloader->PrepareTrainingAndTestTree(
+      mycuts, mycutb, "nTrain_Signal=1000:nTrain_Background=1000:SplitMode=Random:NormMode=NumEvents:!V");
+
+   // Input Layout
+   TString inputLayoutString("InputLayout=1|1|4");
+
+   // Batch Layout
+   TString batchLayoutString("BatchLayout=256|1|4");
+
+   // General layout.
+   TString layoutString("Layout=RESHAPE|1|1|4|FLAT,DENSE|128|TANH,DENSE|128|TANH,DENSE|128|TANH,DENSE|1|LINEAR");
+
+   // Training strategies.
+   TString training0("LearningRate=1e-2,Optimizer=" + optimizerStr +
+                     ",Momentum=0.9,Repetitions=1,"
+                     "ConvergenceSteps=20,BatchSize=256,TestRepetitions=10,"
+                     "WeightDecay=1e-4,Regularization=L2,"
+                     "DropConfig=0.0+0.5+0.5+0.5, Multithreading=True");
```

Comment:
There is an error in the string definition here. You cannot have a whitespace in the string. 
Also the Multithreading option can be removed because it is not activated. 
However this is fixed by my PR #2219. 
I would merge  the PR anyway and then I will fix that when fixing the conflicts I will get in #2219